consensus/dbft/payload.go

```
@@ -61,38 +64,112 @@ func (p *Payload) SetType(t payload.MessageType) {
 
 // Payload implements the payload.ConsensusPayload interface.
 func (p Payload) Payload() any {
-	return p.payload
+	if msg := p.msgCache.Load(); msg != nil {
+		return msg
+	}
+
+	switch p.message.Type {
+	case changeViewType:
+		return p.GetChangeView()
+	case prepareRequestType:
+		return p.GetPrepareRequest()
+	case prepareResponseType:
+		return p.GetPrepareResponse()
+	case commitType:
+		return p.GetCommit()
+	case recoveryRequestType:
+		return p.GetPrepareRequest()
+	case recoveryMessageType:
+		return p.GetRecoveryMessage()
+
+	default:
+		err := fmt.Errorf("invalid type: 0x%02x", byte(p.message.Type))
+		log.Warn(err.Error())
+		return nil
+	}
 }
 
 // SetPayload implements the payload.ConsensusPayload interface.
 func (p *Payload) SetPayload(pl any) {
-	p.payload = pl.(io.Serializable)
+	bytes, err := rlp.EncodeToBytes(pl)
+	if err != nil {
+		panic(fmt.Errorf("newPayload msg rlp encode error: %w", err))
+	}
+	p.MsgPayload = bytes
+	p.msgCache.Store(pl)
 }
 
 // GetChangeView implements the payload.ConsensusPayload interface.
-func (p Payload) GetChangeView() payload.ChangeView { return p.payload.(payload.ChangeView) }
+func (p Payload) GetChangeView() payload.ChangeView {
+	if msg := p.msgCache.Load(); msg != nil {
+		return msg.(*changeView)
+	}
+
+	obj := &changeView{}
+	rlp.DecodeBytes(p.MsgPayload, obj)
+	obj.newViewNumber = p.message.ViewNumber + 1
+	p.msgCache.Store(obj)
+	return obj
+}
```

OK, and now I see that decoding errors are not handled in any way. Could we create a special RLP decoder/encoder for `message` structure so that `MsgPayload` decoding happens immediately after `message` decoding instead of during the first call `GetChangeView` or other getters? And then in this custom RLP `message` decoder we can properly set `msgPayloadCache`.

And in the custom `message` RLP encoder we can serialize `msgPayloadCache` into `MsgPayload` before further standard RLP encoding (if `MsgPayload` is empty).

---

Custom RLP decoder/encoder with `messageExt` type now.